### ISSUE TITLE
Microsoft.Data.SqlClient.SqlException (0x80131904): The statement terminated. The maximum recursion 100 has been exhausted before statement completion.

### ISSUE THREAD

**Issue by EsQueEl-Fella** [trimble-oss/dba-dash#320]
Here another small issue I just discovered:

BTW, Nevermind the bad chart values. I have been (and still am) struggeling with the developers to do something about similar cases. It is a wonder that the big blue circle is not covering the whole Chart! :D. Looking at the situation from positive point of view, if there where no so many blockings, I could not file this exception. ;D

![maximum_recursion](https://user-images.githubusercontent.com/40255900/191992633-e3c6710a-6b11-42d1-b1f6-4a14de9eb69a.png)

************** Ausnahmetext **************
Microsoft.Data.SqlClient.SqlException (0x80131904): The statement terminated. The maximum recursion 100 has been exhausted before statement completion.
   at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at Microsoft.Data.SqlClient.SqlDataReader.TryHasMoreRows(Boolean& moreRows)
   at Microsoft.Data.SqlClient.SqlDataReader.TryReadInternal(Boolean setTimeout, Boolean& more)
   at Microsoft.Data.SqlClient.SqlDataReader.Read()
   at System.Data.Common.DataAdapter.FillLoadDataRow(SchemaMapping mapping)
   at System.Data.Common.DataAdapter.FillFromReader(DataSet dataset, DataTable datatable, String srcTable, DataReaderContainer dataReader, Int32 startRecord, Int32 maxRecords, DataColumn parentChapterColumn, Object parentChapterValue)
   at System.Data.Common.DataAdapter.Fill(DataTable[] dataTables, IDataReader dataReader, Int32 startRecord, Int32 maxRecords)
   at System.Data.Common.DbDataAdapter.FillInternal(DataSet dataset, DataTable[] datatables, Int32 startRecord, Int32 maxRecords, String srcTable, IDbCommand command, CommandBehavior behavior)
   at System.Data.Common.DbDataAdapter.Fill(DataTable[] dataTables, Int32 startRecord, Int32 maxRecords, IDbCommand command, CommandBehavior behavior)
   at System.Data.Common.DbDataAdapter.Fill(DataTable dataTable)
   at DBADashGUI.Performance.RunningQueries.runningQueriesSnapshot(DateTime& snapshotDate, Int32 skip) in D:\a\dba-dash\dba-dash\DBADashGUI\Performance\RunningQueries.cs:line 282
   at DBADashGUI.Performance.RunningQueries.loadSnapshot(DateTime snapshotDate, Int32 skip) in D:\a\dba-dash\dba-dash\DBADashGUI\Performance\RunningQueries.cs:line 299
   at DBADashGUI.Performance.RunningQueries.RefreshData() in D:\a\dba-dash\dba-dash\DBADashGUI\Performance\RunningQueries.cs:line 148
   at DBADashGUI.Performance.RunningQueriesViewer.RunningQueriesViewer_Load(Object sender, EventArgs e) in D:\a\dba-dash\dba-dash\DBADashGUI\Performance\RunningQueriesViewer.cs:line 60
   at System.Windows.Forms.Form.OnLoad(EventArgs e)
   at System.Windows.Forms.Control.CreateControl(Boolean fIgnoreVisible)
   at System.Windows.Forms.Control.CreateControl()
   at System.Windows.Forms.Control.WmShowWindow(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ScrollableControl.WndProc(Message& m)
   at System.Windows.Forms.Form.WmShowWindow(Message& m)
   at System.Windows.Forms.Form.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)
ClientConnectionId:299b14a9-644d-4e06-92ec-5330660cf912
Error Number:530,State:1,Class:16

**Comment by DavidWiseman** [comment]
I can reproduce this by creating a circular reference in the blocking chain - a deadlock. These situations should be resolved by the deadlock monitor but you might occasionally catch these in the running queries snapshots.  I'll see if I can fix it now I'm able to reproduce it.

**Comment by DavidWiseman** [comment]
Issue found due to re-creating the bug.  This issue occurs if wait_time is NULL and query is blocked - this situation shouldn't occur and is down to manual manipulation of the data.  Still worth fixing just in case.
> System.InvalidCastException: Object cannot be cast from DBNull to other types.
>    at System.DBNull.System.IConvertible.ToInt64(IFormatProvider provider)
>    at System.Convert.ToInt64(Object value)
>    at DBADashGUI.Performance.RunningQueries.<>c.<getCounts>b__26_3(DataRow r) in C:\Git\TrimbleOSS\dba-dash\DBADashGUI\Performance\RunningQueries.cs:line 340
>    at System.Linq.Enumerable.Sum[TSource](IEnumerable`1 source, Func`2 selector)
>    at DBADashGUI.Performance.RunningQueries.getCounts() in C:\Git\TrimbleOSS\dba-dash\DBADashGUI\Performance\RunningQueries.cs:line 340
>    at DBADashGUI.Performance.RunningQueries.loadSnapshot(DateTime snapshotDate, Int32 skip) in C:\Git\TrimbleOSS\dba-dash\DBADashGUI\Performance\RunningQueries.cs:line 301
>    at DBADashGUI.Performance.RunningQueries.RefreshData() in C:\Git\TrimbleOSS\dba-dash\DBADashGUI\Performance\RunningQueries.cs:line 149
>    at DBADashGUI.Performance.RunningQueriesViewer.RunningQueriesViewer_Load(Object sender, EventArgs e) in C:\Git\TrimbleOSS\dba-dash\DBADashGUI\Performance\RunningQueriesViewer.cs:line 60
>    at System.Windows.Forms.Form.OnLoad(EventArgs e)
>    at System.Windows.Forms.Form.OnCreateControl()
>    at System.Windows.Forms.Control.CreateControl(Boolean fIgnoreVisible)
>    at System.Windows.Forms.Control.CreateControl()
>    at System.Windows.Forms.Control.WmShowWindow(Message& m)
>    at System.Windows.Forms.Control.WndProc(Message& m)
>    at System.Windows.Forms.ScrollableControl.WndProc(Message& m)
>    at System.Windows.Forms.Form.WmShowWindow(Message& m)
>    at System.Windows.Forms.Form.WndProc(Message& m)
>    at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
>    at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)

**Comment by DavidWiseman** [comment]
Fix is coming soon.  This can be applied manually to the repository DB as a hotfix:

```sql
CREATE OR ALTER FUNCTION dbo.RunningQueriesBlockingHierarchy(
	@InstanceID INT,
	@SnapshotDateUTC DATETIME2(7),
	@BlockingSessionID SMALLINT
)
RETURNS TABLE
AS
RETURN
/* 
	Get the blocking hierarchy for a blocked session starting with the current blocking session  
	For example session 123 might be blocked by session 64.  Session 64 is blocked by session 72.  Session 72 is blocked by session 96 (root blocker)
	We pass in 64 as the current blocking session and get: 96 \ 72 \ 64.  
*/
WITH  H AS(
	SELECT	@BlockingSessionID as blocking_session_id,
			0 as Sort,
			' \ ' + CAST(@BlockingSessionID AS VARCHAR(MAX)) AS BlockingHierarchy,
			0 AS IsDeadlock
	WHERE @BlockingSessionID<>0
	UNION ALL
	SELECT	Q.blocking_session_id,
			H.Sort+1,
			CASE WHEN calc.IsDeadlock =1 THEN ' \ {!Deadlock!}' ELSE '' END + ' \ ' + CAST(Q.blocking_session_id AS VARCHAR(MAX)) + H.BlockingHierarchy,
			calc.IsDeadlock
	FROM H 
	JOIN dbo.RunningQueries Q ON H.blocking_session_id = Q.session_id
	-- Check for a loop (deadlock).  Break execution on the next iteration when a deadlock is detected
	OUTER APPLY(SELECT CASE WHEN Q.blocking_session_id = @BlockingSessionID OR H.BlockingHierarchy LIKE '%\ ' + CAST(Q.blocking_session_id AS VARCHAR(MAX)) + ' \%' THEN 1 ELSE 0 END AS IsDeadlock) calc
	WHERE Q.InstanceID = @InstanceID
	AND Q.SnapshotDateUTC = @SnapshotDateUTC
	AND Q.blocking_session_id <> 0
	AND H.IsDeadlock=0 -- Stop if a deadlock was detected on the previous loop
	AND H.Sort < 100 -- Stop before we break max recursion
	)
SELECT TOP(1) ISNULL(STUFF(BlockingHierarchy,1,3,''),'')  as BlockingHierarchy
FROM H
ORDER BY Sort DESC
```

**Comment by EsQueEl-Fella** [comment]
Hi,
Thx for the fix. I just updated the application to the last version (2.22.4) and applied the fix on top. Things are working correctly now.

Do I have to close the issue or will you do that?

![maximum_recursion_ok](https://user-images.githubusercontent.com/40255900/192082624-b7bb6257-e263-4219-ad3a-10ac77b49699.png)

**Comment by DavidWiseman** [comment]
That's great - thanks for letting me know it's working.  I'll close this case off once I create a new release next week.
An open case that is marked complete is fixed and pending to go out in the next release.

**Comment by DavidWiseman** [comment]
The fix is included in 2.22.5 🎉
